File: LAPBlogsBackEnd/LAPBlogsBackEnd/blog_api/blogs.py
from django.http import JsonResponse,HttpResponse
from .tools import jsonStr
from bson.objectid import ObjectId
import datetime
import pymongo

mongo_client = pymongo.MongoClient(host="127.0.0.1",port=27017)
LAPBlogs = mongo_client["lap_blogs"]

Blogs = LAPBlogs['blogs']
Users = LAPBlogs["users"]

#保存暂存区
def temporarySave(request):
    if request.POST:
        try:
            currentTime = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            user = Users.find_one(ObjectId(request.POST["belongTo"]))
            condition = {"_id":ObjectId(request.POST["belongTo"])}
            user = Users.find_one(ObjectId(request.POST["belongTo"]))
            user["temporarySave"] = {
                "title":request.POST["title"],
                "content":request.POST["content"],
                "saveTime":currentTime
            }
            result = Users.update_one(condition, {'$set': user})
            obj = {"message":"ok","status":1}
        except BaseException as e:
            print(e)
            obj = {"message":"服务器出错","status":-2}
    else:
        obj = {"message":"请求方式出错","status":-5}
    return JsonResponse(obj,safe = False)

#用户的所有博客
def userBlogs(request):
    if request.POST:
        try:
            userBlogsId = request.POST
            blogArray =[]
            for blogId in userBlogsId:
                blog = Blogs.find_one(ObjectId(userBlogsId[blogId]))
                blogArray.append({
                    "blogId":blogId,
                    "detail":blog
                })
            obj = {"message":"请求成功!","status":1,"blogArray":jsonStr(blogArray)}
        except BaseException as e:
            obj = {"message":"服务器出错","status":-2}
    else:
        obj = {"message":"请求方式出错","status":-5}
    return JsonResponse(obj,safe = False)

#用户的单一博客
def userBlog(request):
    if request.POST:
        try:
            userBlogsId = request.POST["userBlogId"]
            blog = Blogs.find_one(ObjectId(userBlogsId))
            obj = {"message":"请求成功!","status":1,"blog":jsonStr(blog)}
        except BaseException as e:
            obj = {"message":"服务器出错","status":-2}
    else:
        obj = {"message":"请求方式出错","status":-5}
    return JsonResponse(obj,safe = False)

#网站请求的所有用户的博客
def requestBlogs(request):
    blogArray = ["Python","JavaScript","NodeJs","Php","Java"]
    BLOGS = {}
    for i in blogArray:
        blogs = Blogs.find({"type":i})
        tempArray = []
        for blog in blogs:
            tempArray.append(blog)
        BLOGS[i] = tempArray
    #BLOGS存储查找到的文章，按照语言类型分类
    BLOGS = jsonStr(BLOGS)
    obj = {
        "message":"success",
        "status":1,
        "blog":BLOGS
    }
    return JsonResponse(obj,safe = False)

#保存博客
def insertBlog(request):
    #用户的_id,用户的name,文章content,文章title,文章type,插入文章
    if request.POST:
        currentTime = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        blog = {
            "belongTo":ObjectId(request.POST["belongTo"]),
            "userName":request.POST["userName"],
            "content":request.POST["content"],
            "title":request.POST["title"],
            "insertTime":currentTime,
            "type":request.POST["type"]
        }
        result = Blogs.insert(blog) #插入blog表返回ObjectId
        user = Users.find_one(ObjectId(request.POST["belongTo"]))
        user["blog"].append(result) #将blog的_id压进user["blog"]
        user["temporarySave"] = {}
        condition = {"_id":ObjectId(request.POST["belongTo"])}
        Users.update(condition,user) #更新user表
        obj = {
            "message":"插入成功",
            "status":1
        }
    else:
        obj = {"message":"请求方式不是post!","status":-5}
    return JsonResponse(obj,safe = False)
        

File: lAPBlogsMiddle/routes/api/blogs.js
const express = require("express");
const router = express.Router();
const md5 = require("crypto-js/md5");
// const qs = require("querystring");
const passport= require("passport");
const request = require("request");

router.post("/test",(req,res)=>{ //测试
    res.json({psth:"/test",env:8010});
});

router.get("/requestBlogs",(req,res)=>{ //请求所有博客
    let url = 'http://127.0.0.1:8015/requestBlogs';
    request.get(url, function(error, httpResponse, body){
        if(error){
            return res.json({"msg":"服务器出错",status:-2});
        }else{
            body = JSON.parse(body);
            if(body.status === 1){
                return res.json({"msg":"success",
                                  "status":1,
                                  "blogs":JSON.parse(body.blog)});
            }
        }
    });
})

router.post("/userBlogs",passport.authenticate("jwt",{session:false}),(req,res)=>{ //用户博客的ids
    let url = 'http://127.0.0.1:8015/userBlogs';
    
    let userBlogs = {
        "userBlogs":req.body.userBlogs
    }
    request.post({url:url, form: userBlogs}, function(err,httpResponse,body){
        body = JSON.parse(body);
        if(body.status === 1){
            res.json({"status":1,"msg":"请求成功!",blogArray:JSON.parse(body.blogArray)});
        }else if(body.status === -5){
            res.json({"status":-5,"msg":"请使用post方式请求!"});
        }else if(body.status === -2){
            res.json({"status":-2,"msg":"服务器出错!"});
        }
    })
})

router.post("/userBlog",passport.authenticate("jwt",{session:false}),(req,res)=>{
    let url = 'http://127.0.0.1:8015/userBlog';
    
    let userBlog = {
        "userBlogId":req.body.blogId
    }
    request.post({url:url, form: userBlog}, function(err,httpResponse,body){
        body = JSON.parse(body);
        if(body.status === 1){
            res.json({"status":1,"msg":"请求成功!",blogArray:JSON.parse(body.blog)});
        }else if(body.status === -5){
            res.json({"status":-5,"msg":"请使用post方式请求!"});
        }else if(body.status === -2){
            res.json({"status":-2,"msg":"服务器出错!"});
        }
    })
})

router.post("/insertBlog",passport.authenticate("jwt",{session:false}),(req,res)=>{//插入博客
    let blog = {
        "belongTo":req.body._id,
        "userName":req.body.userName,
        "title":req.body.title,
        "content":req.body.content,
        "type":req.body.type
    };
    let url = 'http://127.0.0.1:8015/insertBlogs';
    request.post({url:url, form: blog}, function(err,httpResponse,body){
        body = JSON.parse(body);
        if(body.status === 1){
            res.json({"status":1,"msg":"上传成功!"});
        }else if(body.status === -5){
            res.json({"status":-5,"msg":"请使用post方式请求!"});
        }        
    })
});

router.post("/temporarySave",passport.authenticate("jwt",{session:false}),(req,res)=>{//暂时保存博客
    let blog = {
        "belongTo":req.body.belongTo,
        "title":req.body.title,
        "content":req.body.content,
    };
    let url = 'http://127.0.0.1:8015/temporarySave';
    request.post({url:url, form: blog}, function(err,httpResponse,body){
        body = JSON.parse(body);
        if(body.status === 1){
            res.json({"status":1,"msg":"保存成功!"});
        }else if(body.status === -5){
            res.json({"status":-5,"msg":"请使用post方式请求!"});
        }else if(body.status === -2){
            res.json({"status":-2,"msg":"服务器出错!"});
        }
    })
});


    
module.exports = router;